### PR TITLE
#89: Implement SkSession class to wrap a real Slack session 

### PR DESCRIPTION
**Pull request by proshin-roman**
PR for #89. 

I've added a new interface SkSession that should be used instead of SlackSession which is closed for decorating. As well I've added an implementation of this interface - RealSkSession that wraps an instance of SlackSession.

I've added two new puzzles:
- integrate SkSession so that it replaces SlackSession in the code, 
- create fake SkSession class that can be used in tests.

### CODE REVIEW

**Comment by 0crat** [comment]
Job [#1117](https://github.com/zerocracy/farm/issues/1117) is now in scope, role is `REV`

**Comment by 0crat** [comment]
This pull request [#1117](https://github.com/zerocracy/farm/issues/1117) is assigned to @carlosmiranda[/z](https://www.0crat.com/u/carlosmiranda), here is [why](https://www.0crat.com/footprint/C3NDPUA8L/1590673244); the budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4); please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @g4s8[/z](https://www.0crat.com/u/g4s8) (the architect) right in this ticket; if you decide that this PR should not be accepted ever, also inform the architect; this [blog post](http://www.yegor256.com/2015/02/09/serious-code-reviewer.html) will help you understand what is expected from a code reviewer; there will be a monetary reward for this job

File: src/main/java/com/zerocracy/radars/slack/SkSession.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+/**
+ * Session in Slack.
+ *
+ * @author Roman Proshin (roman@proshin.org)
+ * @version $Id$
+ * @since 0.23
+ * @todo #89:30min Use this interface and its implementation RealSkSession
+ *  everywhere instead of SlackSession.
+ */
+public interface SkSession extends AutoCloseable {
```
Comment (carlosmiranda) [inline src/main/java/com/zerocracy/radars/slack/SkSession.java]: @proshin-roman let's extend `Closeable` instead. That already defines `close() throws IOException` so there is no need to redefine `close()` here.

File: src/test/java/com/zerocracy/radars/slack/RealSkSessionTest.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    public void channelUsesOrigin() {
+        final String id = "CHNNL";
+        final SlackSession session = Mockito.mock(SlackSession.class);
+        final SlackChannel channel = Mockito.mock(SlackChannel.class);
+        Mockito.when(session.findChannelById(id))
+            .thenReturn(channel);
+        MatcherAssert.assertThat(
+            "Session returns incorrect channel by its ID",
+            new RealSkSession(session).channel(id),
+            new IsEqual<>(channel)
```
Comment (carlosmiranda) [inline src/test/java/com/zerocracy/radars/slack/RealSkSessionTest.java]: @proshin-roman let's use `Matchers.is(channel)` here.

File: src/test/resources/com/zerocracy/bundles/notify_in_slack_without_user/_before.groovy [hunk 3]
```diff
@@ -20,20 +20,17 @@ import com.jcabi.xml.XML
...
 import org.mockito.Mockito
 
 /**
- * @todo #62:30min We have to implement simple
- *  slack session object to use inside 'notify_in_slack'
- *  stakeholder-script and fake object for tests.
- *  This session object may wrap `SlackSession` from
- *  `com.ullink.slack.simpleslackapi`.
- *  After that we should get rid of `Mockito` in this test.
+ * @todo # 89:30min Add a new class FakeSkSession that implements interface
```
Comment (carlosmiranda) [inline src/test/resources/com/zerocracy/bundles/notify_in_slack_without_user/_before.groovy]: @proshin-roman incorrect formatting of puzzle here. Also, let's call it `FkSkSession` for consistency with existing fake types.

File: src/main/java/com/zerocracy/radars/slack/SkSession.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+import com.ullink.slack.simpleslackapi.SlackUser;
+import java.io.IOException;
+
+/**
+ * Session in Slack.
+ *
+ * @author Roman Proshin (roman@proshin.org)
+ * @version $Id$
+ * @since 0.23
+ * @todo #89:30min Use this interface and its implementation RealSkSession
```
Comment (carlosmiranda) [inline src/main/java/com/zerocracy/radars/slack/SkSession.java]: @proshin-roman this puzzle is too short, PDD will complain. Let's add details (e.g. where is SlackSession used and need to be replaced?)

File: src/test/java/com/zerocracy/radars/slack/RealSkSessionTest.java (hunk 2 above)
Comment (proshin-roman) [reply src/test/java/com/zerocracy/radars/slack/RealSkSessionTest.java]: @carlosmiranda why do you think a static method is better than a clean object? In Cactoos library using of `IsEqual` is more preferred usually. I think it makes sense even to farm project.

**Comment by proshin-roman** [comment]
@carlosmiranda I've fixed your comments - please look at the PR again. As well I wrote an answer for one of your comments.

**Comment by proshin-roman** [comment]
@carlosmiranda by some reason Travis-ci doesn't build this pull request anymore, but I've configured a build for my fork and there this pull request is green: https://travis-ci.org/proshin-roman/farm/builds/390149073

**Comment by g4s8** [comment]
@rultor merge

**Comment by 0crat** [comment]
@ypshenychka[/z](https://www.0crat.com/u/ypshenychka) please review this job completed by @carlosmiranda[/z](https://www.0crat.com/u/carlosmiranda), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by 0crat** [comment]
The job [#1117](https://github.com/zerocracy/farm/issues/1117) is now out of scope

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 point(s) just awarded to @g4s8[/z](https://www.0crat.com/u/g4s8)

**Comment by ypshenychka** [comment]
@0crat quality good

**Comment by 0crat** [comment]
Order was finished, quality is "good": +20 point(s) just awarded to @carlosmiranda[/z](https://www.0crat.com/u/carlosmiranda)

**Comment by 0crat** [comment]
Quality review completed: +8 point(s) just awarded to @ypshenychka[/z](https://www.0crat.com/u/ypshenychka)

**Pull request merged**